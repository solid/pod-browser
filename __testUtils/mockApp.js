/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createSolidDataset,
  createThing,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { vcard, rdf, schema } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { vcardExtras } from "../src/addressBook";

export const oidcPrefix = "http://www.w3.org/ns/solid/oidc#";
export const TOS_PREDICATE = `${oidcPrefix}tos_uri`;
export const POLICY_PREDICATE = `${oidcPrefix}policy_uri`;
export const CONTACTS_PREDICATE = `${oidcPrefix}contacts`;
export const LOGO_PREDICATE = `${oidcPrefix}logo_uri`;

export const APP_WEBID = "https://mockappurl.com";
export const APP_POLICY_URL = "https://mockappurl.com/privacy-policy";
export const APP_TOS_URL = "https://mockappurl.com/terms-of-service";

const contactThing = chain(
  createThing({ url: "https://mockappurl.com#contacts" }),
  (t) => addStringNoLocale(t, vcard.fn, "Example Contact"),
  (t) => addStringNoLocale(t, vcard.hasTelephone, "555-5555"),
  (t) => addStringNoLocale(t, vcard.hasEmail, "example@email.com"),
  (t) => addUrl(t, vcard.url, "https://examplewebsite.com")
);

const dataset = setThing(createSolidDataset(), contactThing);

export function mockApp() {
  return chain(
    createThing({ url: APP_WEBID }),
    (t) => addStringNoLocale(t, vcard.fn, "Mock App"),
    (t) => addUrl(t, vcardExtras("WebId"), APP_WEBID),
    (t) => addUrl(t, rdf.type, schema.SoftwareApplication), // just guessing what the type might be for apps
    (t) => addUrl(t, TOS_PREDICATE, APP_TOS_URL),
    (t) => addUrl(t, POLICY_PREDICATE, APP_POLICY_URL),
    (t) => addUrl(t, CONTACTS_PREDICATE, "https://mockappurl.com#contacts")
  );
}

export function mockAppDataset() {
  return setThing(dataset, mockApp());
}
