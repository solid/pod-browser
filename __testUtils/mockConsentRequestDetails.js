/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

async function getConsentRequestDetails() {
  return {
    "@context": [
      "https://www.w3.org/2018/credentials/v1",
      "https://w3id.org/security/suites/ed25519-2020/v1",
      "https://consent.pod.inrupt.com/credentials/v1",
    ],
    id: "https://consent.pod.inrupt.com/vc/53973727-f4d0-9e8dbdc041fd",
    type: ["VerifiableCredential", "SolidCredential", "SolidConsentRequest"],
    issuer: "https://consent.pod.inrupt.com",
    issuanceDate: "2021-05-26T16:40:03Z",
    expirationDate: "2021-06-23T16:40:03Z",
    credentialSubject: {
      id: "https://pod.inrupt.com/bob/profile/card#me",
      inbox: "https://pod.inrupt.com/bob/inbox/",
      hasConsent: {
        mode: ["Read", "Write"],
        hasStatus: "ConsentStatusRequested",
        forPersonalData: "https://pod.inrupt.com/alice/private/data",
        forPurpose: "https://example.com/SomeSpecificPurpose",
      },
    },
    proof: {
      created: "2021-05-26T16:40:03.009Z",
      proofPurpose: "assertionMethod",
      proofValue: "eqp8h_kL1DwJCpn65z-d1Arnysx6b11...jb8j0MxUCc1uDQ",
      type: "Ed25519Signature2020",
      verificationMethod: "https://consent.pod.inrupt.com/key/396f686b4ec",
    },
  };
}

export default getConsentRequestDetails;
