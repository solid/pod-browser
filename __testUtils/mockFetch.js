/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

export default function mockFetch(responses = {}) {
  return jest.fn().mockImplementation((url, init) => {
    const responseFn = responses[url.toString()];
    const responseObject = responseFn;
    const method = init?.method || "GET";
    let response;
    if (responseFn && typeof responseObject[method] !== "undefined") {
      response = responseObject[method](url, init);
    } else if (responseFn) {
      response = responseFn(url, init);
    } else {
      throw new Error(`URL (${url}) not mocked properly`);
    }
    Object.defineProperty(response, "url", { value: url });
    return Promise.resolve(response);
  });
}

export function mockSolidClientRequest(body, method = "GET", headers = {}) {
  return new Request(body, {
    headers: new Headers(headers),
    method,
  });
}

export function mockTurtleResponse(status, body = "") {
  return new Response(body, {
    status,
    headers: {
      "Content-Type": "text/turtle",
    },
  });
}
