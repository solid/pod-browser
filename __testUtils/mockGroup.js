/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  asUrl,
  getThing,
  mockSolidDatasetFrom,
  mockThingFrom,
  removeUrl,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { foaf, rdf, vcard } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { getBaseUrl } from "../src/solidClientHelpers/resource";

export function addMembersToMockedGroup(group, agentUrls) {
  const dataset = chain(
    group.dataset,
    ...agentUrls.map((agentUrl) => (d) =>
      setThing(d, setUrl(group.thing, vcard.hasMember, agentUrl))
    )
  );

  return {
    dataset,
    thing: getThing(dataset, asUrl(group.thing)),
  };
}

export function removeMembersFromMockedGroup(group, agentUrls) {
  const dataset = chain(
    group.dataset,
    ...agentUrls.map((agentUrl) => (d) =>
      setThing(d, removeUrl(group.thing, vcard.hasMember, agentUrl))
    )
  );

  return {
    dataset,
    thing: getThing(dataset, asUrl(group.thing)),
  };
}

export function mockGroupThing(name, url, { description, members } = {}) {
  return chain(
    mockThingFrom(url),
    (t) => setUrl(t, rdf.type, vcard.Group),
    (t) => setStringNoLocale(t, foaf.name, name),
    ...(members || []).map((agentUrl) => (t) =>
      addUrl(t, vcard.hasMember, agentUrl)
    ),
    ...[
      description ? (t) => setStringNoLocale(t, vcard.note, description) : null,
    ].filter((fn) => !!fn)
  );
}

export default function mockGroup(name, url, options = {}) {
  const thing = mockGroupThing(name, url, options);
  const dataset = setThing(mockSolidDatasetFrom(getBaseUrl(url)), thing);
  return { dataset, thing };
}
