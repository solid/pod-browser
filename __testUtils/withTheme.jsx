/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { render } from "@testing-library/react";
import T from "prop-types";
import { StylesProvider, ThemeProvider } from "@inrupt/prism-react-components";
import { FeatureProvider } from "../src/contexts/featureFlagsContext";
import defaultTheme from "../src/theme";

export default function WithTheme(props) {
  const { theme, children } = props;
  return (
    <StylesProvider generateClassName={(rule) => `PodBrowser-${rule.key}`}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </StylesProvider>
  );
}

WithTheme.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  theme: T.object,
  children: T.node.isRequired,
};

WithTheme.defaultProps = {
  theme: defaultTheme,
};

export const renderWithTheme = (children, theme = defaultTheme) => {
  return render(
    <FeatureProvider>
      <WithTheme theme={theme}>{children}</WithTheme>
    </FeatureProvider>
  );
};
