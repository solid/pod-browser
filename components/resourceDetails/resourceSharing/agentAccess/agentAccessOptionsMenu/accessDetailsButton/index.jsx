/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* eslint-disable react/jsx-one-expression-per-line */

import React, { useState, useEffect } from "react";
import T from "prop-types";
import { createStyles, ListItem, ListItemText } from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import { permission as permissionPropType } from "../../../../../../constants/propTypes";
import styles from "./styles";

export const TESTCAFE_ID_VIEW_DETAILS_BUTTON = "view-details-button";
const useStyles = makeStyles((theme) => createStyles(styles(theme)));

export default function AccessDetailsButton({ setOpenModal }) {
  const classes = useStyles();

  return (
    <ListItem
      data-testid={TESTCAFE_ID_VIEW_DETAILS_BUTTON}
      button
      onClick={() => setOpenModal(true)}
    >
      <ListItemText
        disableTypography
        classes={{ primary: classes.listItemText }}
      >
        View Details
      </ListItemText>
    </ListItem>
  );
}

AccessDetailsButton.propTypes = {
  setOpenModal: T.func.isRequired,
};
