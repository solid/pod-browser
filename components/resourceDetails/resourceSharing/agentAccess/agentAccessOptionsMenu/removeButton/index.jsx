/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useContext, useEffect, useState } from "react";
import PropTypes from "prop-types";
import { createStyles, ListItem, ListItemText } from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import { getResourceName } from "../../../../../../src/solidClientHelpers/resource";
import AccessControlContext from "../../../../../../src/contexts/accessControlContext";
import ConfirmationDialogContext from "../../../../../../src/contexts/confirmationDialogContext";
import styles from "./styles";
import { PUBLIC_AGENT_PREDICATE } from "../../../../../../src/models/contact/public";
import { AUTHENTICATED_AGENT_PREDICATE } from "../../../../../../src/models/contact/authenticated";
import {
  permission as permissionPropType,
  profile as profilePropType,
} from "../../../../../../constants/propTypes";
import ResourceInfoContext from "../../../../../../src/contexts/resourceInfoContext";
import ConfirmationDialog from "../../../../../confirmationDialog";

export const handleRemovePermissions = ({
  setLoading,
  accessControl,
  setLocalAccess,
  mutateResourceInfo,
}) => {
  return async (agentWebId, policyName) => {
    setLoading(true);
    if (PUBLIC_AGENT_PREDICATE === agentWebId) {
      accessControl.setRulePublic(policyName, false);
    }
    if (AUTHENTICATED_AGENT_PREDICATE === agentWebId) {
      accessControl.setRuleAuthenticated(policyName, false);
    }
    const { response: updatedAcr } = await accessControl.removeAgentFromPolicy(
      agentWebId,
      policyName
    );
    await mutateResourceInfo(updatedAcr, false);
    setLoading(false);
    setLocalAccess(null);
  };
};

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

export const TESTCAFE_ID_REMOVE_BUTTON = "remove-button";
export const DIALOG_ID = "remove-agent";
export const CONFIRMATION_TEXT = "Remove";

export default function RemoveButton({
  resourceIri,
  permission: { webId, alias },
  profile,
  setLoading,
  setLocalAccess,
}) {
  const { accessControl } = useContext(AccessControlContext);
  const { mutate: mutateResourceInfo } = useContext(ResourceInfoContext);
  const resourceName = getResourceName(resourceIri);
  const classes = useStyles();
  const [bypassDialog, setBypassDialog] = useState(false);
  const [confirmationSetup, setConfirmationSetup] = useState(false);
  const handleRemoveAgent = handleRemovePermissions({
    setLoading,
    accessControl,
    setLocalAccess,
    mutateResourceInfo,
  });

  const {
    confirmed,
    open,
    setConfirmed,
    setOpen,
    title,
    setTitle,
    setConfirmText,
    closeDialog,
  } = useContext(ConfirmationDialogContext);

  const handleOpenDialog = () => {
    setConfirmText(CONFIRMATION_TEXT);
    // eslint-disable-next-line prettier/prettier
    if (
      PUBLIC_AGENT_PREDICATE === webId ||
      AUTHENTICATED_AGENT_PREDICATE === webId
    ) {
      setBypassDialog(true);
    }
    const text = `Remove ${
      profile?.name || webId
    }'s access from ${resourceName}`;
    setTitle(text);
    setOpen(DIALOG_ID);
  };

  useEffect(() => {
    setConfirmationSetup(true);
    if (bypassDialog) {
      setConfirmed(true);
      handleRemoveAgent(webId, alias);
    }
    if (open !== DIALOG_ID) return;
    if (confirmationSetup && confirmed === null) return;

    if (confirmationSetup && confirmed) {
      handleRemoveAgent(webId, alias);
    }

    if (confirmationSetup && confirmed !== null) {
      closeDialog();
      setConfirmationSetup(false);
    }
    setConfirmText(null);
  }, [
    confirmationSetup,
    confirmed,
    title,
    webId,
    alias,
    bypassDialog,
    setConfirmed,
    setConfirmationSetup,
    setConfirmText,
    handleRemoveAgent,
    open,
    closeDialog,
  ]);

  return (
    <ListItem
      data-testid={TESTCAFE_ID_REMOVE_BUTTON}
      button
      onClick={handleOpenDialog}
    >
      <ListItemText
        disableTypography
        classes={{ primary: classes.listItemText }}
      >
        Remove
      </ListItemText>
      <ConfirmationDialog />
    </ListItem>
  );
}

RemoveButton.propTypes = {
  resourceIri: PropTypes.string.isRequired,
  permission: permissionPropType.isRequired,
  profile: profilePropType,
  setLoading: PropTypes.func,
  setLocalAccess: PropTypes.func,
};

RemoveButton.defaultProps = {
  setLoading: () => {},
  setLocalAccess: () => {},
  profile: null,
};
