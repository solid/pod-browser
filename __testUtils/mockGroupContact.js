/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  asUrl,
  getThing,
  mockSolidDatasetFrom,
  mockThingFrom,
  removeThing,
  removeUrl,
  setThing,
} from "@inrupt/solid-client";
import { chain } from "../src/solidClientHelpers/utils";
import { createGroupDatasetUrl } from "../src/models/contact/group";
import { vcardExtras } from "../src/addressBook";
import { mockGroupThing } from "./mockGroup";
import { getAddressBookDatasetUrl } from "../src/models/addressBook";

export function mockIndexThing(indexDataset, addressBook, groupThingUrl) {
  const addressBookThingUrl = asUrl(
    addressBook.thing,
    getAddressBookDatasetUrl(addressBook)
  );
  const addressBookThing =
    getThing(indexDataset, addressBookThingUrl) ||
    mockThingFrom(addressBookThingUrl);
  return addUrl(addressBookThing, vcardExtras("includesGroup"), groupThingUrl);
}

export function addGroupToMockedIndexDataset(
  dataset,
  addressBook,
  name,
  groupThingUrl
) {
  return chain(
    dataset,
    (d) => setThing(d, mockGroupThing(name, groupThingUrl)),
    (d) => setThing(d, mockIndexThing(d, addressBook, groupThingUrl))
  );
}

export function removeGroupFromMockedIndexDataset(
  dataset,
  addressBook,
  groupIndex,
  groupThing,
  groupThingUrl
) {
  return chain(
    dataset,
    (d) => removeThing(d, groupThing),
    (d) =>
      setThing(
        d,
        chain(
          getThing(groupIndex, asUrl(addressBook.thing)),
          (t) => removeUrl(t, vcardExtras("includesGroup"), groupThingUrl) // remove the includesGroup reference to group
        )
      )
  );
}

export default function mockGroupContact({
  addressBook,
  name,
  indexUrl = null,
  options = {},
}) {
  const { url, id } = options;
  const groupDatasetUrl = url || createGroupDatasetUrl(addressBook, id);
  const groupThingUrl = `${groupDatasetUrl}#this`;
  const groupThing = mockGroupThing(name, groupThingUrl, options);
  return {
    dataset: chain(
      mockSolidDatasetFrom(indexUrl || groupDatasetUrl),
      (d) => setThing(d, groupThing),
      (d) => setThing(d, mockIndexThing(d, addressBook, groupThingUrl))
    ),
    thing: groupThing,
  };
}
