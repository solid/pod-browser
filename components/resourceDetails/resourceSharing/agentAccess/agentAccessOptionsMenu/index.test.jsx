/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";
import { screen, waitFor } from "@testing-library/dom";
import * as accesGrantFns from "@inrupt/solid-client-access-grants";
import { renderWithTheme } from "../../../../../__testUtils/withTheme";
import AgentAccessOptionsMenu from "./index";
import { TESTCAFE_ID_REMOVE_BUTTON } from "./removeButton";
import getSignedVc from "../../../../../__testUtils/mockSignedVc";

import { TESTCAFE_ID_VIEW_DETAILS_BUTTON } from "./accessDetailsButton";
import {
  TESTCAFE_ID_ACCESS_DETAILS_DONE_BUTTON,
  TESTCAFE_ID_ACCESS_DETAILS_MODAL,
  TESTCAFE_ID_ACCESS_DETAILS_REVOKE_BUTTON,
} from "./accessDetailsButton/accessDetailsModal";

const resourceIri = "/iri/";
const webId = "https://example.com/profile/card#me";
const profile = {
  avatar: null,
  name: "Example Agent",
  webId,
};

describe("AgentAccessOptionsMenu", () => {
  it("renders a button which triggers the opening of the menu", () => {
    const permission = { webId, profile, alias: "editors" };
    const { asFragment, getByTestId, queryByText } = renderWithTheme(
      <AgentAccessOptionsMenu
        resourceIri={resourceIri}
        permission={permission}
        setLoading={jest.fn}
        setLocalAccess={jest.fn}
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(queryByText("WebId:")).toBeNull();
    const menuButton = getByTestId("menu-button");
    userEvent.click(menuButton);
    expect(queryByText("WebId:")).toBeDefined();
    const removeButton = getByTestId(TESTCAFE_ID_REMOVE_BUTTON);
    expect(removeButton).toBeDefined();
  });
});

describe("Access Details Modal", () => {
  it("renders a details modal when you click on the view details button", async () => {
    const permission = {
      webId,
      alias: "Editors",
      type: "agent",
      vc: getSignedVc(),
    };
    const { asFragment, getByTestId, queryByText } = renderWithTheme(
      <AgentAccessOptionsMenu
        resourceIri={resourceIri}
        permission={permission}
        setLoading={jest.fn}
        setLocalAccess={jest.fn}
      />
    );

    const menuButton = getByTestId("menu-button");
    userEvent.click(menuButton);
    await waitFor(() => {
      const viewDetailsButton = getByTestId(TESTCAFE_ID_VIEW_DETAILS_BUTTON);
      expect(viewDetailsButton).toBeInTheDocument();
      userEvent.click(viewDetailsButton);
    });

    await waitFor(() => {
      expect(getByTestId(TESTCAFE_ID_ACCESS_DETAILS_MODAL)).toBeInTheDocument();
    });
  });

  it("closes the details modal when you click on the revoke button", async () => {
    const permission = {
      webId,
      alias: "Editors",
      type: "agent",
      vc: getSignedVc(),
    };
    jest.spyOn(accesGrantFns, "revokeAccessGrant").mockResolvedValue(true);

    const { getByTestId } = renderWithTheme(
      <AgentAccessOptionsMenu
        resourceIri={resourceIri}
        permission={permission}
        setLoading={jest.fn}
        setLocalAccess={jest.fn}
      />
    );

    const menuButton = getByTestId("menu-button");
    userEvent.click(menuButton);

    await waitFor(() => {
      const viewDetailsButton = getByTestId(TESTCAFE_ID_VIEW_DETAILS_BUTTON);
      userEvent.click(viewDetailsButton);
    });
    const revokeButton = getByTestId(TESTCAFE_ID_ACCESS_DETAILS_REVOKE_BUTTON);
    userEvent.click(revokeButton);

    userEvent.click(revokeButton);
    await waitFor(() => {
      const closedModal = screen.queryByTestId(
        TESTCAFE_ID_ACCESS_DETAILS_MODAL
      );
      expect(closedModal).toBeNull();
    });
  });

  it("closes the details modal when you click on the done button", async () => {
    const permission = {
      webId,
      alias: "Editors",
      type: "agent",
      vc: getSignedVc(),
    };

    const { getByTestId } = renderWithTheme(
      <AgentAccessOptionsMenu
        resourceIri={resourceIri}
        permission={permission}
        setLoading={jest.fn}
        setLocalAccess={jest.fn}
      />
    );

    const menuButton = getByTestId("menu-button");
    userEvent.click(menuButton);

    await waitFor(() => {
      const viewDetailsButton = getByTestId(TESTCAFE_ID_VIEW_DETAILS_BUTTON);
      userEvent.click(viewDetailsButton);
    });
    const revokeButton = getByTestId(TESTCAFE_ID_ACCESS_DETAILS_DONE_BUTTON);
    userEvent.click(revokeButton);

    act(() => userEvent.click(revokeButton));
    await waitFor(() => {
      const closedModal = screen.queryByTestId(
        TESTCAFE_ID_ACCESS_DETAILS_MODAL
      );
      expect(closedModal).toBeNull();
    });
  });
});
