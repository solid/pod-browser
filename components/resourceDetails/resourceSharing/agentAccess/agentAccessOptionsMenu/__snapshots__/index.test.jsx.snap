// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgentAccessOptionsMenu renders a button which triggers the opening of the menu 1`] = `
<DocumentFragment>
  <button
    class="PodBrowser-button"
    data-testid="menu-button"
    type="button"
  >
    <i
      class="PodBrowser-icon-more"
    />
  </button>
</DocumentFragment>
`;
